### ISSUE TITLE
Possible Bug With ASTReader.Read

### ISSUE THREAD

**Issue by ryanfalzon** [ConsenSys/solc-typed-ast#99]
## Environment

I have a scenario where I am currently undergoing analyses of a solidity AST within a .NET Core application. I have successfully managed to create an environment within a .NET Core application to successfully compile and execute TypeScript code from within the application itself

My end goal is the following:

- From within a .NET Core console application invoke a script written in TypeScript to compile a solidity file and return the AST back to the console application.
- Carry out necessary analysis and if needed any changes  to the AST using .NET Core.
- Pass the updated AST to another script written in TypeScript to turn the AST back to solidity code.

TypeScript Version: 4.4

## Scenario

Running the following script:

```typescript
import { CompileFailedError, ASTReader, compileSol } from "solc-typed-ast";

var generateAst = async function (callback: (error: any, result: any) => void, filepath: string) {
    try {
        const result = await compileSol(filepath, "auto", []);

        const reader = new ASTReader();
        const sourceUnits = reader.read(result.data);

        callback(null, sourceUnits);
    } catch (e) {
        let errorMessage = e.Message;

        if (e instanceof CompileFailedError) {
            errorMessage = "Compile errors encountered:\n";

            for (const failure of e.failures) {
                errorMessage += `\nSolc ${failure.compilerVersion}:`;

                for (const error of failure.errors) {
                    errorMessage += `\n${error}`;
                }
            }
        }

        callback({
            success: false,
            error: errorMessage
        }, null);
    }
};

export = generateAst;
```

Where the contents of the Solidity file are:

```javascript
pragma solidity ^0.5.0;

contract TestExpressionStatement {
    function single(int x) public {
        1 + 2;
    }

    function multiple() public {
        int x = 0;

        (int a, int b) = (1, 2);

        (b, a) = (a, b);

        x = x + a - b;

        x > 10 || x == 10;
    }

    function nested() public {
        int x;

        for (x = 0; x < 10; x++) {
            x += 2;
        }

        x++;
    }
}
```

## Current behavior

The following error is produced:

```
TypeError: Converting circular structure to JSON
    --> starting at object with constructor 'SourceUnit'
    |     property 'ownChildren' -> object with constructor 'Array'
    |     index 0 -> object with constructor 'PragmaDirective'
    --- property 'parent' closes the circle
```

**Comment by ryanfalzon** [comment]
For reference the generated .js file that is created after compiling the TypeScript script is the following:

```javascript
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var solc_typed_ast_1 = require("solc-typed-ast");
var generateAst = function (callback, filepath) {
    return __awaiter(this, void 0, void 0, function () {
        var result, reader, sourceUnits, e_1, errorMessage, _i, _a, failure, _b, _c, error;
        return __generator(this, function (_d) {
            switch (_d.label) {
                case 0:
                    _d.trys.push([0, 2, , 3]);
                    return [4 /*yield*/, (0, solc_typed_ast_1.compileSol)(filepath, "auto", [])];
                case 1:
                    result = _d.sent();
                    reader = new solc_typed_ast_1.ASTReader();
                    sourceUnits = reader.read(result.data);
                    callback(null, sourceUnits);
                    return [3 /*break*/, 3];
                case 2:
                    e_1 = _d.sent();
                    errorMessage = e_1.Message;
                    if (e_1 instanceof solc_typed_ast_1.CompileFailedError) {
                        errorMessage = "Compile errors encountered:\n";
                        for (_i = 0, _a = e_1.failures; _i < _a.length; _i++) {
                            failure = _a[_i];
                            errorMessage += "\nSolc ".concat(failure.compilerVersion, ":");
                            for (_b = 0, _c = failure.errors; _b < _c.length; _b++) {
                                error = _c[_b];
                                errorMessage += "\n".concat(error);
                            }
                        }
                    }
                    callback({
                        success: false,
                        error: errorMessage
                    }, null);
                    return [3 /*break*/, 3];
                case 3: return [2 /*return*/];
            }
        });
    });
};
module.exports = generateAst;
//# sourceMappingURL=generateAst.js.map
```

**Comment by blitz-1306** [comment]
@ryanfalzon Hello, Ryan.

The output of `ASTReader.read()` is normalized AST that **intentionally** contains circular references (to provide referenced nodes resolition and advanced tree traversal logic). If passed `callback` will try to convert it to JSON, then you will encounter the error you mentioning.

Could you investigate in the callback you passing is trying to apply `JSON.encode()` to `sourceUnits`?

**Comment by ryanfalzon** [comment]
@blitz-1306 Spot on! It was in fact the case that the callback automatically tries to convert the result to Json. If anyone has the same issue this was the solution:

```typescript
const getCircularReplacer = () => {
    const seen = new WeakSet();
    return (key, value) => {
        if (typeof value === "object" && value !== null) {
            if (seen.has(value)) {
                return;
            }
            seen.add(value);
        }
        return value;
    };
};
```
and then use that function as the second parameter to create the JSON manually:
```typescript
const result = await compileSol(filepath, "auto", []);

const reader = new ASTReader();
const sourceUnits = reader.read(result.data);
const json = JSON.stringify(sourceUnits, getCircularReplacer());
```

**Comment by ryanfalzon** [comment]
@blitz-1306 another quick question, it is my understanding that the ```writer.Write()``` function takes the AST generated from the ```reader.Read()``` function and gives me back Solidity code. Do you know why the following code just gives me back the sourceUnits?

```typescript
var generateCode = async function (callback: (error: any, result: any) => void, sourceUnits: SourceUnit[]) {
    try {
        const formatter = new PrettyFormatter(4, 0);
        const writer = new ASTWriter(
            DefaultASTWriterMapping,
            formatter,
            LatestCompilerVersion
        );

        let result = "";
        for (const sourceUnit of sourceUnits) {
            result += writer.write(sourceUnit);
        }
        
        callback(null, result);
    }
    catch (e) {
        callback(null, e.message);
    }
};
```

**Comment by blitz-1306** [comment]
@ryanfalzon Well, I may have an idea why this may happen.I would suggest _to avoid_ trying to encode `sourceUnits` and pass as JSON due to they are _intended_ to have circular references. By removing references you also may cause issues with the functionality that relies on `parent`, `children` and so on. AST writer also may rely on this logic.

What you may consider to do insteadis passing raw AST from compiler, that is stored in `result.data` property. Then construct `ASTReader` a bit more later in your logic and call `ASTReader.read()` in place where you actually need to process normalized and typed AST and you need extended functionality (analysis, traversal, writing).

In short:
- `generateAst()` should proparate `result.data` instead of trying to encode source units (and also damage them due to replacing circular references).
- `generateCode()` should constrcut an `ASTReader` and `ASTReader.read()` get "correct source units", then process them somehow, and at last construct `ASTWriter` and apply `ASTWriter.write()`.

**Comment by blitz-1306** [comment]
@ryanfalzon Hello there. Are there any updates on the issue? Or, shoult it be closed?

**Comment by blitz-1306** [comment]
I'm closing this issue due to there are no updates. Hope that discussion helped.
